param.go

```diff
@@ -146,10 +146,35 @@ func (pl paramList) Build(containerStore) (reflect.Value, error) {
 // to the underlying constructor.
 func (pl paramList) BuildList(c containerStore) ([]reflect.Value, error) {
 	args := make([]reflect.Value, len(pl.Params))
+	argsBuilt := make([]bool, len(pl.Params))
+	allContainers := c.getStoresFromRoot()
 	for i, p := range pl.Params {
 		var err error
-		args[i], err = p.Build(c)
-		if err != nil {
+		var arg reflect.Value
+		// iterate through the tree path of scopes.
+		for _, c := range allContainers {
+			if arg, err = p.Build(c); err == nil {
+				args[i] = arg
+				argsBuilt[i] = true
+			}
+		}
+
+		// If an argument failed to build, that means none of the
+		// scopes had the type. This should be reported.
+		if !argsBuilt[i] {
+			return nil, err
+		}
+
+		// If argument has successfully been built, it's possible
+		// for these errors to occur in child scopes that don't
+		// contain the given parameter type. We can safely ignore
+		// these.
+		// If it's an error other than missing types/dependencies,
+		// this means some constructor returned an error that must
+		// be reported.
+		_, isErrMissingTypes := err.(errMissingTypes)
+		_, isErrMissingDeps := err.(errMissingDependencies)
+		if err != nil && !isErrMissingTypes && !isErrMissingDeps {
```

Comment:
I'm not sure I understand what's going here fully:
`err` is assigned in a loop above. The value of `err` here will be the last thing assigned to it before the loop exits.
So this will report an error only if the a constructor in the middle somewhere failed, but one of the ones after it succeeded.

---

Is there risk here in implementing the scope-resolution logic at such a high level in the parameter tree?
If we implemented it on the leaf nodes (paramSingle and maybe paramGroupedSlice), we might be able to control how these operate in a multi-scope environment. Like, paramSingle.Build could call getValueProviders on current containerStore, then its parent, and so on until it finds a provider or runs out of stores.